docs/pkgtorrent-service.8
.TH pkgtorrent-service 8 "09 MARCH 2016" ""
.SH NAME
.B pkgtorrent-service(8)
Is a python WSGI web application that can generate BitTorrent files for binary files available from the web server.
.SH SYNOPSIS
The program should installed on a site local
.B yum
mirror and must be configured before use. See \fBpkgtorrent-service.conf(5)\fR for more information.
.SH USAGE
The primary use case for this service is implemented in \fByum-torrent(1)\fR.
.PP
There is a secondary interface to assist with simple file downloads involving a single path.
.SH EXAMPLE
.RS
.nf
$ curl http://example.com/\fBpath/to/large/file\fR > \fBfile\fR
.fi
.RE
.PP
can be converted to using BitTorrent with:
.PP
.RS
.nf
$ curl http://example.com/\fBpath\fR/torrent_service/single/\fBpath/to/large/file\fR > \fBfile\fR.torrent
$ btclient \fBfile\fR.torrent
.fi
.RE
.PP
Note the repetition of \fBpath\fR before the service name, and the addition of \fBtorrent_service/single\fR to pass control to the single path mode.
Simple file mode supports simple globbing, so you can specify ...\fB/directory/*.foo\fR.
If the path resolves to only one file, the multi-file support in the .torrent file is skipped, making the downloaded file easier to handle.
.SH INPUT FILES
.IP /var/lib/torrent_service/trackers
The service needs to include one or more BitTorrent tracker servers.
The format is one "announce" URL per line. Example:
.B http://example.com:6969/announce
This text file is scanned each minute for updates.
.SH OUTPUT FILES
The service makes extensive use of cache files. The files are relatively small. It is recommended you periodically clean these caches for files not used in the past month or so.
.IP /var/cache/torrent_service/info/\fBinput_digest\fR
This directory caches \fIinfo_hashes\fR for the previously requested path digests.
These are used with the current list of trackers to generate .torrent files in constant time.
.IP /var/cache/torrent_service/pieces/path/to/file/\fIblocksize\fR
Every file within a torrent is represented by a list of SHA1 digests, one for each \fIblocksize\fR of the file.
The SHA1 list is expensive to generate.
The same binaries may be downloaded in different permutations by different clients.
The cache is invalidated if the original file has a different mtime or is a different size.
.SH BUGS
Single file mode should be called simple.
.SH AUTHOR
Matthew Almond <malmond@fb.com>
.SH SEE ALSO
.BR pkgtorrent-service.conf(5)


ChangeLog
2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* yum-utils.spec: bumped yum-utils version to 1.1.31

2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* ChangeLog: updated ChangeLog

2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* yum-utils.spec: bumped yum-utils version to 1.1.31

2011-08-03  Shawn Starr <shawn.starr@rogers.com>

	* docs/yum-utils.1: Update yum-utils man page, to add new utils.

2011-07-25  James Antill <james@and.org>

	Merge branch 'master' of
	ssh://yum.baseurl.org/srv/projects/yum/git/yum-utils  * 'master' of
	ssh://yum.baseurl.org/srv/projects/yum/git/yum-utils: (3 commits)
	typo leaving .conf in ybc.conf.yumvar should just be ybc.yumvar
	...

2011-07-25  James Antill <james@and.org>

	* plugins/aliases/aliases: Add some new aliases, fix chkD.

2011-07-25  Seth Vidal <skvidal@fedoraproject.org>

	* yum-config-manager.py: typo leaving .conf in ybc.conf.yumvar
	should just be ybc.yumvar

2011-07-12  James Antill <james@and.org>

	* plugins/security/security.py: Also auto-apply --security when
	given --sec-severity for other commands.

2011-07-08  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Add repoquery @group completion.

2011-07-08  Ville Skyttä <ville.skytta@iki.fi>

	* repoquery.py: Fix "nitmes" typo in previous commit.

2011-07-06  James Antill <james@and.org>

	* repoquery.py: Fix typo

2011-06-29  James Antill <james@and.org>

	* repoquery.py: Allow @foo groups as a packages selector in
	repoquery.

2011-06-25  Ville Skyttä <ville.skytta@iki.fi>

	* plugins/aliases/aliases.py, plugins/security/security.py,
	plugins/versionlock/versionlock.py: Remove some unused local
	variables and imports flagged by pyflakes.

2011-06-24  Ville Skyttä <ville.skytta@iki.fi>

	* debuginfo-install.py, package-cleanup.py, yum-builddep.py,
	yum-complete-transaction.py: Check doUtilBuildTransaction return
	value, exit early on failure.  BZ 716267

2011-06-22  James Antill <james@and.org>

	* plugins/security/security.py: Try to be clever when sorting the
	security severities.

2011-06-17  Raman Gupta <raman@rocketraman.com>

	* plugins/merge-conf/merge-conf.py: After (external) merge is
	complete, give user option to delete other file

2011-06-17  Raman Gupta <raman@rocketraman.com>

	* plugins/merge-conf/merge-conf.py: Support meld as well as vimdiff
	for merging  When using meld, the left side has the "other file"
	i.e. the file with the .rpmnew or .rpmsave extension. The right side
	has the "final file" i.e. the target file into which changes have to
	be merged.  If the rpm action was to save the existing conf file,
	then the left side will be the original conf file (.rpmsave) and the
	right side will be the new rpm-based configuration. The user should
	merge any changes they want to keep from their old .rpmsave from
	left to right.  If the rpm action was to keep the existing conf
	file, then the left side will be the new conf file (.rpmnew) and
	right side will be the current local configuration. The user should
	merge any changes in the new rpm file into their local config from
	left to right.  Ideally these operations would be 3-way diffs
	between the local file and the new rpm config, using the original
	rpm config as the common merge base. However, the original rpm
	config is not available to yum and therefore a 3-way diff is not
	possible. Perhaps a future enhancement would be to use a .orig file,
	if present, as the merge base for meld.  When using vimdiff, the
	order has not changed. The first parameter to vimdiff is the "final:
	file i.e. the target file, and the second parameter is the "other"
	file i.e. the .rpmnew or .rpmsave file.

2011-06-10  James Antill <james@and.org>

	* repodiff.py: Huge speedup for large number of add/remove obsoletes
	processing.

2011-06-09  Nick Jacek <njacek@redhat.com>

	* yumdownloader.py: Stops yumdownloader from attempting to download
	the same package twice if it is available in multiple repos. BZ
	711767

2011-06-10  James Antill <james@and.org>

	* docs/repodiff.1, repodiff.py: Add --compare-arch for the other
	side of the multilib. diff. BZ 710579.

2011-06-09  Nick Jacek <njacek@redhat.com>

	* repodiff.py: Changed archlist set up to handle multilib. BZ 710579


2011-06-01  James Antill <james@and.org>

	* docs/repoquery.1: Add releasever and setopt docs. from yum to
	repoquery. BZ 638542.

2011-05-20  Seth Vidal <skvidal@fedoraproject.org>

	* yum-utils.spec: frelling pyyaml pkg named with weird case

2011-05-20  Seth Vidal <skvidal@fedoraproject.org>

	* plugins/puppetverify/puppetverify.conf,
	plugins/puppetverify/puppetverify.py, yum-utils.spec: add
	puppetverify plugin

2011-05-18  Frank Weltersbach <frankwe@live.com>

	* docs/package-cleanup.1: Fix typo in package-cleanup man page. BZ
	705617.

2011-05-17  Seth Vidal <skvidal@fedoraproject.org>

	* repoquery.py:     - fix rh bug: #705179     - fix fmt_queryformat not being happy
	    with new 'output' formats passed       everywhere else.

2011-04-26  James Antill <james@and.org>

	* yumdownloader.py: Allow files/provides to be used in
	yumdownloader.

2011-04-26  James Antill <james@and.org>

	* yumdownloader.py: Exit with code 2, if we can't download the
	packages asked for. BZ 699469.

2011-04-26  James Antill <james@and.org>

	* plugins/auto-update-debuginfo/auto-update-debuginfo.py: try/except
	the open, for the rpmdbv cached data (a-u-d plugin).

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Add yum-debug-dump completion.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-debug-dump.py: Mention filename arg in yum-debug-dump usage
	message.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-debug-dump.py: Make yum-debug-dump.py executable for running
	in-place.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Use functions from recent bash-completion if
	available.

2011-04-06  James Antill <james@and.org>

	* repoquery.py, yum-groups-manager.py: Convert

plugins/pkgtorrent/server/pkgtorrent-service.py
#
# Copyright (c) 2014-present, Facebook, Inc.
# All rights reserved.
#
# This source code is licensed under the GPLv2 license found in the COPYING
# file in the root directory of this source tree.
#
import os
import cgi
import cPickle as pickle
import time
import datetime
import locale
import glob

try:
    from hashlib import sha1
except ImportError:
    # fall back.
    from sha import new as sha1

# expires headers must be in en_US names, so set this once.
locale.setlocale(locale.LC_TIME, 'en_US')

# derived from https://gist.github.com/stendec/949007
# original doc text:
#   Bencoding
#   By Stendec <stendec365@gmail.com>
#   Released under the BSD license. Do whatever.
def bencode(o):
    if isinstance(o, unicode):
        o = o.encode('utf8')
    if isinstance(o, str):
        return '%d:%s' % (len(o), o)
    elif type(o) in (list, set, tuple):
        return 'l%se' % ''.join([bencode(x) for x in o])
    elif type(o) in (int, long):
        return 'i%se' % (o,)
    elif isinstance(o, dict):
        return 'd%se' % ''.join(
            ['%s%s' % (bencode(k), bencode(o[k])) for k in sorted(o.keys())]
        )

    raise TypeError('Cannot encode object of type, %s' % str(type(object)))

# if the pickle datastructures change, bump this to invalidate caches.
PROTOCOL = 4
CACHEROOT = '/var/cache/torrent_service'
PIECECACHE = os.path.join(CACHEROOT, 'pieces')
INFOCACHE = os.path.join(CACHEROOT, 'info')
INFOCACHEMAX = 3600  # number of seconds a given torrent is valid for
INFOCACHEWAIT = 25   # how long to wait for another process to create the info
TRACKERS_LIST_PATH = '/var/lib/torrent_service/trackers'
# number of seconds between stat'ing TRACKERS_LIST_PATH for change
TRACKERS_LIST_INTERVAL = 60
NULLPATH = '/nullfile/'
BLOCKSIZE = 2 ** 20  # 1MB should be enough for anybody.

if not os.path.exists(PIECECACHE):
    os.makedirs(PIECECACHE)
if not os.path.exists(INFOCACHE):
    os.makedirs(INFOCACHE)

def create_nullfile(path_info, start_response):
    size = path_info[len(NULLPATH):].split('_')[0]

    if int(size) > BLOCKSIZE:
        out = 'Refusing to create a nullfile bigger than server BLOCKSIZE'
        start_response('403 Forbidden', [
            ('Content-type', 'text/plain'),
            ('Content-Length', str(len(out))),
        ])
        return [out]

    start_response('200 OK', [
        ('Content-type', 'application/octet-stream'),
        ('Content-Transfer-Encoding', 'binary'),
        ('Content-Disposition', 'attachment;'),
        (
            'Expires',  # Not convinced this helps with caching
            (
                datetime.datetime.utcnow() + datetime.timedelta(days=1)
            ).strftime('%a, %d %b %Y %H:%M:%S GMT'),
        ),
        ('Content-Length', size),
    ])
    return ['\0' * int(size)]

# it expired at the epoch to start with.
trackers, trackers_expires, trackers_mtime = None, 0, 0
def update_trackers():
    global trackers, trackers_expires, trackers_mtime
    now = time.time()
    if trackers_expires > now:
        return
    trackers_expires = now + TRACKERS_LIST_INTERVAL
    if os.stat(TRACKERS_LIST_PATH).st_mtime == trackers_mtime:
        return
    trackers = open(TRACKERS_LIST_PATH).read().splitlines()

def create_torrent(environ, start_response):
    update_trackers()
    form = cgi.FieldStorage(
        fp=environ['wsgi.input'],
        environ=environ,
    )
    # support 3 different path specs:
    # /digest/{digest}
    # /single/path/to/file
    path_info = environ['PATH_INFO']
    # but let's make digest work first.
    if path_info.startswith('/digest/'):
        digest = os.path.basename(path_info[8:])
        pathspecs = form.getlist('p')
        single_mode = False
    elif path_info.startswith('/single/'):
        # pathspecs start with the leading / from single
        pathspecs = [path_info[7:]]
        digest = sha1('\n'.join(pathspecs)).hexdigest()
        single_mode = True
    log = environ['wsgi.errors']
    document_root = environ['DOCUMENT_ROOT']
    document_root_len = len(document_root)
    service_name = environ['SCRIPT_NAME'].split('/')[-1]

    paths = []
    if pathspecs:
        # build paths from possible wildcards
        for pathspec in pathspecs:
            if pathspec.endswith(os.pathsep + '**'):
                # rsync style, directory and all subdirectories
                raise NotImplementedError('You gotta write this')
            for p in glob.glob(os.path.join(document_root, pathspec[1:])):
                s = os.stat(p)
                paths.append((p[document_root_len:], s.st_mtime))
    single = single_mode and len(paths) == 1

    digestfile = os.path.join(INFOCACHE, digest)
    # look up a previously created torrent
    try:
        # single file torrents aren't padded and do not have leading directories
        # so should be cached using a slightly different filename
        if single:
            digestfile += '_s'
        digeststat = os.stat(digestfile)
        if digeststat.st_mtime < (time.time() - INFOCACHEMAX):
            # file is too old, could be changed
            os.unlink(digestfile)
            raise KeyError('Cached torrent is too old')

        if digeststat.st_size == 0:
            # another process should be creating this.
            # let's take the mtime of the file, add in INFOCACHEWAIT and
            # watch it for changes.
            waituntil = digeststat.st_mtime + INFOCACHEWAIT
            while True:
                # yuck, polling.
                time.sleep(1)
                digeststat = os.stat(digestfile)
                if digeststat.st_size > 0:
                    # yea, something got written, so let's use it. Note we
                    # expect the file to get replaced atomically (which we do)
                    break
                if time.time() > waituntil:
                    raise KeyError('Timed out waiting for other writer')

        data = pickle.load(open(digestfile))
        if data[0] != PROTOCOL:
            os.unlink(digestfile)
            raise KeyError('Cache protocol is too old')

        protocol, info, info_hash, cached_paths, url_list_suffix = data
        if paths and cached_paths != paths:
            raise KeyError('Path from cache do not match, rebuild')
        return return_torrent(
            environ,
            start_response,
            info,
            info_hash,
            url_list_suffix,
        )
    except (OSError, IOError, KeyError), e:
        log.write(str(e))
        if not pathspecs:
            # Would use 404 here, but our caching infra caches 404s
            # and totally ignores all the controls for it.
            start_response('412 Precondition failed', [
                ('Cache-Control', 'no-cache, no-store, must-revalidate'),
                ('Pragma', 'no-cache'),
                ('Expires', '0'),
            ])
            return []

    # if we get this far, we should have a list of pathspecs to compare
    # to the expected digest.
    assert sha1('\n'.join(pathspecs)).hexdigest() == digest

    # It's possible we get a storm of people suddenly asking for exactly the
    # same set of files, so make them block and return the cached one.
    open(digestfile, 'w').close()
    files = []
    pieces = []
    lookup = []
    n = 0
    prefixes = set()
    nullfile = None
    nullusage = {}
    for path, mtime in paths:
        # these always start with / as they're relative to the root
        path = path[1:]
        cachepath = os.path.join(PIECECACHE, path, str(BLOCKSIZE))
        splitpath = path.split(os.sep)
        prefixes.add(splitpath[0])

        valid_cache = False
        padding = None
        if os.path.exists(cachepath):
            try:
                data = pickle.load(open(cachepath))
                if data[0] == PROTOCOL:
                    protocol, size, blocks, tail, cached_mtime = data
                    valid_cache = protocol == PROTOCOL and cached_mtime == mtime
                    padding = BLOCKSIZE - (size % BLOCKSIZE)
                else:
                    valid_cache = False
            except NotImplementedError:
                # TODO: figure out pickle and file exceptions
                valid_cache = False
        if not valid_cache:
            fullpath = os.path.join(document_root, path)
            size, blocks, tail = os.stat(fullpath).st_size, [], None
            f = open(fullpath)
            for i in xrange(int(size / BLOCKSIZE)):
                m = sha1()
                lookup.append((len(files), f.tell()))
                m.update(f.read(BLOCKSIZE))
                blocks.append(m.digest())
            remainder = size % BLOCKSIZE
            if remainder:
                m = sha1()
                # read as much as we can left
                lookup.append((len(files), f.tell()))
                buffer = f.read()
                assert len(buffer) == remainder
                m.update(buffer)
                tail = m.digest()
                padding = BLOCKSIZE - remainder
                m.update('\0' * padding)
                blocks.append(m.digest())
            blocks = ''.join(blocks)
            f.close()
            cachedir = os.path.dirname(cachepath)
            if not os.path.exists(cachedir):
                os.makedirs(cachedir)
            pickle.dump(
                (PROTOCOL, size, blocks, tail, mtime),
                open(cachepath, 'wb'),
                pickle.HIGHEST_PROTOCOL,
            )
        # url-list requires a name, and that needs stripped
        # off all the paths inside the torrent to work.
        files.append({
            'path': splitpath[1:],
            'length': size,
        })
        if padding:
            n = nullusage.setdefault(padding, 0)
            name = '%d_%d' % (padding, n)
            nullusage[padding] = n + 1
            nullfile = {
                'path': [service_name, 'nullfile', name],
                'length': padding,
            }
            files.append(nullfile)
        pieces.append(''.join(blocks))

    try:
        assert len(prefixes) == 1
    except:
        log.write(repr(pieces))
        raise

    if files[-1] == nullfile:
        # we keep track of the last nullfile path spec we created in the loop.
        # the last file in the torrent doesn't need to be null padded.
        del files[-1]
        pieces[-1] = pieces[-1][:-len(tail)] + tail

    if single:
        url_list_suffix, name = os.path.split(path)
        info = {
            'name': name,
            'length': size,
        }
    else:
        url_list_suffix = ''
        info = {
            'name': list(prefixes)[0],
            'files': files,
        }
    info.update({
        'piece length': BLOCKSIZE,
        'pieces': ''.join(pieces),
    })
    info_hash = sha1(bencode(info)).hexdigest()

    tmpfile = '%s.tmp' % (digestfile,)
    pickle.dump(
        (PROTOCOL, info, info_hash, paths, url_list_suffix),
        open(tmpfile, 'wb'),
        pickle.HIGHEST_PROTOCOL
    )
    os.rename(tmpfile, digestfile)

    return return_torrent(
        environ,
        start_response,
        info,
        info_hash,
        url_list_suffix
    )

def return_torrent(environ, start_response, info, info_hash, url_list_suffix):
    # we want a list of trackers to put in the announce list that is salted by
    # the info_hash, so different info_hashes get distributed around different
    # primary trackers.
    announce = dict((sha1(t + info_hash).digest(), t) for t in trackers)
    announce_list = [[announce[k]] for k in sorted(announce.keys())]

    torrent = bencode({
        'url-list': [
            'http://%s/%s/' % (environ['HTTP_HOST'], url_list_suffix),
        ],
        'announce-list': announce_list,
        'info': info,
    })

    headers = [
        ('Content-type', 'application/x-bittorrent'),
        ('Content-Transfer-Encoding', 'binary'),
        ('Content-Disposition', 'attachment; filename=%s.torrent' % info_hash,),
        ('Content-Length', str(len(torrent))),
    ]
    start_response('200 OK', headers)
    return [torrent]

def application(environ, start_response):
    path_info = environ['PATH_INFO']
    if path_info.startswith(NULLPATH):
        return create_nullfile(path_info, start_response)
    return create_torrent(environ, start_response)
